sermant-plugins/sermant-router/router-common/src/main/java/io/sermant/router/common/xds/XdsRouterHandler.java

```diff
@@ -0,0 +1,231 @@
+    }
+
+    private Set<ServiceInstance> handleXdsRoute(XdsRoute route, String serviceName) {
+        // select cluster
+        XdsRouteAction routeAction = route.getRouteAction();
+        String cluster = routeAction.getCluster();
+        if (routeAction.isWeighted()) {
+            cluster = selectClusterByWeight(routeAction.getWeightedClusters());
+        }
+
+        // get service instance of cluster
+        Optional<XdsClusterLoadAssigment> loadAssigmentOptional =
+                serviceDiscovery.getClusterServiceInstance(cluster);
+        if (!loadAssigmentOptional.isPresent()) {
+            return serviceDiscovery.getServiceInstance(serviceName);
+        }
+        XdsClusterLoadAssigment clusterLoadAssigment = loadAssigmentOptional.get();
+
+        if (!routeService.isLocalityRoute(clusterLoadAssigment.getClusterName())) {
+            Set<ServiceInstance> serviceInstances = getServiceInstanceOfCluster(clusterLoadAssigment);
+            return serviceInstances.isEmpty() ? serviceDiscovery.getServiceInstance(serviceName) : serviceInstances;
+        }
+
+        // get locality info of self-service and route by locality
+        Optional<XdsLocality> localityInfoOfSelfService = XdsRouterUtil.getLocalityInfoOfSelfService();
+        if (localityInfoOfSelfService.isPresent()) {
+            Set<ServiceInstance> serviceInstances = getServiceInstanceOfLocalityCluster(clusterLoadAssigment,
+                    localityInfoOfSelfService.get());
+            if (!serviceInstances.isEmpty()) {
+                return serviceInstances;
+            }
+        }
+
+        Set<ServiceInstance> serviceInstances = getServiceInstanceOfCluster(clusterLoadAssigment);
+        return serviceInstances.isEmpty() ? serviceDiscovery.getServiceInstance(serviceName) : serviceInstances;
+    }
+
+    private Set<ServiceInstance> getServiceInstanceOfLocalityCluster(XdsClusterLoadAssigment clusterLoadAssigment,
+            XdsLocality locality) {
+        return clusterLoadAssigment.getLocalityInstances().entrySet().stream()
+                .filter(xdsLocalitySetEntry -> xdsLocalitySetEntry.getKey().equals(locality))
+                .flatMap(xdsLocalitySetEntry -> xdsLocalitySetEntry.getValue().stream())
+                .collect(Collectors.toSet());
+    }
+
+    private Set<ServiceInstance> getServiceInstanceOfCluster(XdsClusterLoadAssigment clusterLoadAssigment) {
+        return clusterLoadAssigment.getLocalityInstances().entrySet().stream()
+                .flatMap(instanceEntry -> instanceEntry.getValue().stream())
+                .collect(Collectors.toSet());
+    }
+
+    private boolean isPathMatched(XdsPathMatcher matcher, String path) {
+        return matcher.isMatch(path);
+    }
+
+    private boolean isHeadersMatched(List<XdsHeaderMatcher> matchers, Map<String, String> headers) {
+        return matchers.stream()
+                .allMatch(xdsHeaderMatcher -> xdsHeaderMatcher.isMatch(headers));
```

Comment:
every Matcher rule should return true？ Or use anyMatch instead of allMatch

Reply:
yes, all header rules must be matched.